src/desktop/tabletinput.cpp

```diff
@@ -2,89 +2,232 @@
 #endif
 
 namespace tabletinput {
 
 static const char *inputMode = "Qt tablet input";
 
-void init(QApplication *app, const QSettings &cfg)
-{
-#if defined(Q_OS_WIN) && defined(HAVE_KIS_TABLET)
-	bool useWindowsInk = false;
-	// Enable Windows Ink tablet event handler
-	// This was taken directly from Krita
-	if(cfg.value("settings/input/windowsink", true).toBool()) {
-		KisTabletSupportWin8 *penFilter = new KisTabletSupportWin8();
-		if(penFilter->init()) {
-			app->installNativeEventFilter(penFilter);
-			useWindowsInk = true;
-			inputMode = "KisTablet Windows Ink input";
-			qDebug("Using Win8 Pointer Input for tablet support");
+#ifdef Q_OS_WIN
+
+static Mode currentMode = Mode::Uninitialized;
+
+class SpontaneousTabletEventFilter final : public QObject {
+public:
+	explicit SpontaneousTabletEventFilter(QObject *parent)
+		: QObject{parent}
+	{
+	}
+
+protected:
+	bool eventFilter(QObject *watched, QEvent *event) override final
+	{
+		switch(event->type()) {
+		case QEvent::TabletEnterProximity:
+		case QEvent::TabletLeaveProximity:
+		case QEvent::TabletMove:
+		case QEvent::TabletPress:
+		case QEvent::TabletRelease:
+			// KisTablet uses QApplication::sendEvent, which means its events
+			// are never set to be spontaneous. We use that to filter out Qt's
+			// own tablet events, which in turn are always spontaneous.
+			if(event->spontaneous()) {
+				return false;
+			}
+			Q_FALLTHROUGH();
+		default:
+			return QObject::eventFilter(watched, event);
+		}
+	}
+};
 
+static std::unique_ptr<KisTabletSupportWin8> kisTabletSupportWin8;
+static std::unique_ptr<SpontaneousTabletEventFilter>
+	spontaneousTabletEventFilter;
+
+static Mode getModeFromSettings(const QSettings &cfg)
+{
+	if(cfg.contains("settings/input/tabletinputmode")) {
+		int value = cfg.value("settings/input/tabletinputmode").toInt();
+		if(value > int(Mode::Uninitialized) && value <= int(Mode::Last)) {
+			return Mode(value);
 		} else {
-			qWarning("No Win8 Pointer Input available");
-			delete penFilter;
+			qWarning("Invalid settings/input/tabletinputmode: %d", value);
 		}
+	}
+	// Drawpile 2.1 had separate Windows Ink and relative pen mode hack
+	// settings, translate them if we don't find the new one.
+	if(cfg.value("settings/input/windowsink", true).toBool()) {
+		return Mode::KisTabletWinink;
+	} else if(cfg.value("settings/input/relativepenhack", false).toBool()) {
+		return Mode::KisTabletWintabRelativePenHack;
+	} else {
+		return Mode::KisTabletWintab;
+	}
+}
+
+Mode extractMode(const QSettings &cfg)
+{
+	// Annoying fallback definitions to get the closest available mode.
+#	if QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
+	constexpr Mode qt5 = Mode::Qt6Winink;
+	constexpr Mode qt6Winink = Mode::Qt6Winink;
+	constexpr Mode qt6Wintab = Mode::Qt6Wintab;
+#	else
+	constexpr Mode qt5 = Mode::Qt5;
+	constexpr Mode qt6Winink = Mode::Qt5;
+	constexpr Mode qt6Wintab = Mode::Qt5;
+#	endif
+
+	Mode mode = getModeFromSettings(cfg);
+	switch(getModeFromSettings(cfg)) {
```

Comment:
making the same call twice here

Reply:
Corrected to use the variable as it should.